compiler/plc_lowering/src/inheritance.rs

```diff
@@ -0,0 +1,1853 @@
+        Self {
+            base_type_name: Some(base_type_name.into()),
+            pou: self.pou.clone(),
+            id_provider: self.id_provider.clone(),
+        }
+    }
+
+    fn with_pou(&self, pou: impl Into<String>) -> Self {
+        Self {
+            base_type_name: self.base_type_name.clone(),
+            pou: Some(pou.into()),
+            id_provider: self.id_provider.clone(),
+        }
+    }
+
+    fn provider(&self) -> IdProvider {
+        self.id_provider.clone()
+    }
+}
+
+pub struct InheritanceLowerer {
+    pub index: Option<Index>,
+    pub annotations: Option<Box<dyn AnnotationMap>>,
+    ctx: Context,
+}
+
+impl InheritanceLowerer {
+    pub fn new(id_provider: IdProvider) -> Self {
+        Self { index: None, annotations: None, ctx: Context::new(id_provider) }
+    }
+
+    pub fn provider(&self) -> IdProvider {
+        self.ctx.provider()
+    }
+
+    pub fn visit_unit(&mut self, unit: &mut CompilationUnit) {
+        self.visit_compilation_unit(unit);
+    }
+
+    fn walk_with_context<T: WalkerMut>(&mut self, t: &mut T, ctx: Context) {
+        let old_ctx = std::mem::replace(&mut self.ctx, ctx);
+        t.walk(self);
+        self.ctx = old_ctx;
+    }
+
+    // Updates the base to reflect the inheritance chain of the node and returns it
+    fn update_inheritance_chain(
+        &self,
+        node: &AstNode,
+        base: Option<Box<AstNode>>,
+        base_type: &str,
+    ) -> Option<Box<AstNode>> {
+        if self.index.is_none() || self.annotations.is_none() {
+            return base;
+        }
+
+        let annotations = self.annotations.as_ref().expect("Annotations exist");
+        let Some(qualified_name) = annotations.get_qualified_name(node) else {
+            return base;
+        };
+
+        let segment = qualified_name.split('.').next().expect("Must have a name");
+
+        if base_type == segment {
+            return base;
+        }
+
+        let index = self.index.as_ref().expect("Index exists");
+        let inheritance_chain = index.get_inheritance_chain(base_type, segment);
+        if inheritance_chain.len() <= 1 {
+            return base;
+        }
+
+        // add a `__BASE` qualifier for each element in the inheritance chain, exluding `self`
+        inheritance_chain.iter().skip(1).fold(base, |base, _| {
+            Some(Box::new(AstFactory::create_member_reference(
+                AstFactory::create_identifier(
+                    BASE_TYPE_NAME,
+                    SourceLocation::internal(),
+                    self.provider().next_id(),
+                ),
+                base.map(|it: Box<AstNode>| *it),
+                self.provider().next_id(),
+            )))
+        })
+    }
+}
+
+impl AstVisitorMut for InheritanceLowerer {
+    fn visit_pou(&mut self, pou: &mut Pou) {
+        if self.index.is_some() {
+            return self.walk_with_context(pou, self.ctx.with_pou(&pou.name));
```

Comment:
I think the `walk_with_context` can be replaced with `pou.walk(self)` here, no? As I understand this the `visit_pou` is only responsible for patching the `__SUPER` variable right after parsing, past that it doesn't have any other tasks other than calling the default walker?

Reply:
Doing so would cause reference-expressions in variable initializers to not be visited in the correct context. I will add a test to showcase this